### PR TITLE
Add docs for codeql workspaces

### PR DESCRIPTION
**Pull request by aeisenberg**
Documentation for the `codeql-workspace.yml` file. What it is, how and why to use it.

### CODE REVIEW

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+In most cases, the CodeQL workspace and all CodeQL packs contained in it should be stored in the same git repository so the development environment is more easily sharable.
+
+The ``codeql-workspae.yml`` file
+--------------------------------
+
+A CodeQL workspace is defined by a ``codeql-workspace.yml`` yaml file. This file contains a ``provide`` block, and optionally an ``ignore`` block. The ``provide`` block contains a list of glob patterns that define the CodeQL packs that are available in the workspace. The ``ignore`` block contains a list of glob patterns that define CodeQL packs that are not available in the workspace. Each entry in the ``provide`` or ``ignore`` section must map to a path to a ``qlpack.yml`` file. All glob patterns are relative to the directory containing the workspace file. See `@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>` for a list of patterns accepted in this file.
+
+For example, the following ``codeql-workspace.yml`` file defines a workspace that contains all CodeQl packs recursively found in the ``codeql-packs`` directory, except for the packs in the ``experimental`` directory:
```
Comment (intrigus-lgtm) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:19]:
```suggestion
For example, the following ``codeql-workspace.yml`` file defines a workspace that contains all CodeQL packs recursively found in the ``codeql-packs`` directory, except for the packs in the ``experimental`` directory:
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
+.. _about-codeql-workspaces:
+
+About CodeQL Workspaces
+=======================
+
+.. include:: ../reusables/beta-note-package-management.rst
+
+CodeQL workspaces are used to group multiple CodeQL packs together. CodeQL packs in the same workspace are automatically available as source dependencies for each other when running any CodeQL command that resolves queries. This makes it easier to develope and maintain multiple, related CodeQL packs. A typical use case for a CodeQL workspace is for developing one or more CodeQL library packs and one or more query packs that depends on it in them in the same location.
```
Comment (intrigus-lgtm) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:8]:
```suggestion
CodeQL workspaces are used to group multiple CodeQL packs together. CodeQL packs in the same workspace are automatically available as source dependencies for each other when running any CodeQL command that resolves queries. This makes it easier to develop and maintain multiple, related CodeQL packs. A typical use case for a CodeQL workspace is for developing one or more CodeQL library packs and one or more query packs that depends on it in them in the same location.
```

> A typical use case for a CodeQL workspace is for developing one or more CodeQL library packs and one or more query packs that depends on it in them in the same location.

I failed to parse this sentence.

Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:8]: Thanks for the review! I will reword this to make it more understandable.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+In most cases, the CodeQL workspace and all CodeQL packs contained in it should be stored in the same git repository so the development environment is more easily sharable.
+
+The ``codeql-workspae.yml`` file
+--------------------------------
+
+A CodeQL workspace is defined by a ``codeql-workspace.yml`` yaml file. This file contains a ``provide`` block, and optionally an ``ignore`` block. The ``provide`` block contains a list of glob patterns that define the CodeQL packs that are available in the workspace. The ``ignore`` block contains a list of glob patterns that define CodeQL packs that are not available in the workspace. Each entry in the ``provide`` or ``ignore`` section must map to a path to a ``qlpack.yml`` file. All glob patterns are relative to the directory containing the workspace file. See `@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>` for a list of patterns accepted in this file.
```
Comment (jf205) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:17]: I think the link is missing two trailing underscores.

```suggestion
A CodeQL workspace is defined by a ``codeql-workspace.yml`` yaml file. This file contains a ``provide`` block, and optionally an ``ignore`` block. The ``provide`` block contains a list of glob patterns that define the CodeQL packs that are available in the workspace. The ``ignore`` block contains a list of glob patterns that define CodeQL packs that are not available in the workspace. Each entry in the ``provide`` or ``ignore`` section must map to a path to a ``qlpack.yml`` file. All glob patterns are relative to the directory containing the workspace file. See `@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>`__ for a list of patterns accepted in this file.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+Similarly, publishing a CodeQL query pack to the GitHub container registry using  ``codeql pack publish`` will always use dependencies found in the workspace instead of using dependencies found in the local package cache.
+
+This ensures that any local change to a query library in a dependency in the same workspace will be automatically reflected in the published query pack.
+
+.. pull-quote::
+
+  Note
+
+  Source dependencies are CodeQL packs that are resolved from the filesystem. They might be in the same CodeQL workspace, or specified a path option in the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and version constraints are ignored. This ensures that during local development version mismatches can be ignored.
```
Comment (jf205) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:44]: Missing 'as'?

```suggestion
  Source dependencies are CodeQL packs that are resolved from the filesystem. They might be in the same CodeQL workspace, or specified as a path option in the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and version constraints are ignored. This ensures that during local development version mismatches can be ignored.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+.. code-block:: yaml
+
+    name: my-company/my-queries
+    version: 1.0.0
+    dependencies:
+      my-company/my-library: "*"
+      codeql/cpp-all: ~0.2.0
+
+Notice that ``"*"`` is specified as the version constraint for the library pack. Because the library pack is a source dependency, the version constraint is not needed since the library pack's content is always resolved from inside of the workspace. Any version constraint will be ignored in this case, but it is recommended to use ``"*"`` for source dependencies to avoid confusion.
```
Comment (jf205) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:74]:
```suggestion
Notice that, for ``my-company/my-queries``, ``"*"`` is specified as the version constraint for the library pack in the ``dependencies`` block. The library pack is defined as a source dependency in ``codeql-workspace.yml``, so the version constraint is not needed since the library pack's content is always resolved from inside of the workspace. Any version constraint will be ignored in this case, but it is recommended to use ``"*"`` for source dependencies to avoid confusion.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+When ``codeql pack install`` is executed from the query pack directory, an appropriate version of ``codeql/cpp-all`` will be downloaded to the local package cache. Also, a ``codeql-pack.lock.yml`` file will be created that contains the resolved version of ``codeql/cpp-all``, but no entry for ``my-company/my-library`` since it is resolved from source. The ``codeql-pack.lock.yml`` file will look something like this:
```
Comment (jf205) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:76]:
```suggestion
When ``codeql pack install`` is executed from the query pack directory, an appropriate version of ``codeql/cpp-all`` will be downloaded to the local package cache. Also, a ``codeql-pack.lock.yml`` file will be created that contains the resolved version of ``codeql/cpp-all``. The lock file won't contain an entry for ``my-company/my-library`` since it is resolved from source. The ``codeql-pack.lock.yml`` file will look something like this:
```

File: docs/codeql/codeql-cli/about-codeql-packs.rst [hunk 7]
```diff
@@ -199,10 +199,8 @@ The ``codeql-pack.lock.yml`` file will contain something like the following:
...
 
-..
-  TODO: Add a link to the CodeQL CLI documentation for query resolution, specifically in regards to resolving from source
 
-The ``codeql/cpp-all`` dependency is locked to version 0.1.4. The ``my-user/my-lib`` dependency is locked to version 0.2.4. The ``my-user/transitive-dependency``, which is a transitive dependency and is not specified in the ``qlpack.yml`` file, is locked to version 1.2.4. The ``other-dependency/from-source`` is absent from the lock file since it is resolved from source. This dependency must be available in the same CodeQL workspace as the pack.
+The ``codeql/cpp-all`` dependency is locked to version 0.1.4. The ``my-user/my-lib`` dependency is locked to version 0.2.4. The ``my-user/transitive-dependency``, which is a transitive dependency and is not specified in the ``qlpack.yml`` file, is locked to version 1.2.4. The ``other-dependency/from-source`` is absent from the lock file since it is resolved from source. This dependency must be available in the same CodeQL workspace as the pack. For more information about CodeQL workspaces and resolving dependencies from source see ":doc:`About CodeQL Workspaces <about-codeql-workspaces>`."
```
Comment (jf205) [inline docs/codeql/codeql-cli/about-codeql-packs.rst:203]:
```suggestion
The ``codeql/cpp-all`` dependency is locked to version 0.1.4. The ``my-user/my-lib`` dependency is locked to version 0.2.4. The ``my-user/transitive-dependency``, which is a transitive dependency and is not specified in the ``qlpack.yml`` file, is locked to version 1.2.4. The ``other-dependency/from-source`` is absent from the lock file since it is resolved from source. This dependency must be available in the same CodeQL workspace as the pack. For more information about CodeQL workspaces and resolving dependencies from source, see ":doc:`About CodeQL Workspaces <about-codeql-workspaces>`."
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst (hunk 3 above)
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:17]: Hmmmm....that doesn't look right. 

About_CodeQL_Workspaces_—_CodeQL

I'll figure out what is going on.

**Review by aeisenberg — COMMENTED** [review]
Thanks for your suggestions.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst (hunk 3 above)
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:17]: User error. Your suggestion was correct. :)

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    ignore:
+      - "*/codeql-packs/**/experimental/**/qlpack.yml"
+
+To verify that you have the correct ``codeql-workspace.yml`` file, run ``codeql pack ls`` command in the same directory as your workspace. The result of the command is a list of all CodeQL packs in the workspace.
+
+
+CodeQL workspaces and query resolution
+--------------------------------------
+
+All CodeQL packs in a workspace are available as source dependencies for each other when running any CodeQL command that resolves queries or packs. For example, when ``codeql pack install`` is run in a pack directory in a workspace, any dependency found in the workspace will not be downloaded to the package cache, nor will it be added to the resulting ``codeql-pack.lock.yml`` file. See `:ref:Adding and Installing Dependencies <adding-and-installing-dependencies>` for more information.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:34]:
```suggestion
All CodeQL packs in a workspace are available as source dependencies for each other when you run any CodeQL command that resolves queries or packs. For example, when you run ``codeql pack install`` in a pack directory in a workspace, any dependency that can be found in the workspace will be instead of downloading that dependency to the package cache and adding it to the ``codeql-pack.lock.yml`` file. For more information, see `:ref:Adding and Installing Dependencies <adding-and-installing-dependencies>`__.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+Similarly, publishing a CodeQL query pack to the GitHub container registry using  ``codeql pack publish`` will always use dependencies found in the workspace instead of using dependencies found in the local package cache.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:36]:
```suggestion
Similarly, when you publish a CodeQL query pack to the GitHub container registry using  ``codeql pack publish`` the command will always use the dependencies from the workspace instead of using dependencies found in the local package cache.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+This ensures that any local change to a query library in a dependency in the same workspace will be automatically reflected in the published query pack.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:38]:
```suggestion
This ensures that any local changes you make to a query library in a dependency are automatically reflected in any query packs you publish from that workspace.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+.. pull-quote::
+
+  Note
+
+  Source dependencies are CodeQL packs that are resolved from the filesystem. They might be in the same CodeQL workspace, or specified as a path option in the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and version constraints are ignored. This ensures that during local development version mismatches can be ignored.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:44]:
```suggestion
  Source dependencies are CodeQL packs that are resolved from the local file system. These dependencies can be in the same CodeQL workspace, or specified as a path option using the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and any version constraints defined in the cache are ignored. This ensures that when you are developing CodeQL locally any version mismatches can be ignored.
```

I've suggested what I think is a simpler construction above, but it's true to say that I'm left wondering if that means that you're likely to publish a CodeQL pack with version mismatches that might cause your users problems.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+.. code-block:: yaml
+
+    name: my-company/my-queries
+    version: 1.0.0
+    dependencies:
+      my-company/my-library: "*"
+      codeql/cpp-all: ~0.2.0
+
+Notice that, for ``my-company/my-queries``, ``"*"`` is specified as the version constraint for the library pack in the ``dependencies`` block. The library pack is defined as a source dependency in ``codeql-workspace.yml``, so the version constraint is not needed since the library pack's content is always resolved from inside of the workspace. Any version constraint will be ignored in this case, but it is recommended to use ``"*"`` for source dependencies to avoid confusion.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:74]:
```suggestion
Notice that the ``dependencies`` block for the CodeQL query pack,``my-company/my-queries``, specifies  ``"*"`` as the version of the library pack. Since the library pack is already defined as a source dependency in ``codeql-workspace.yml``, the library pack's content is always resolved from inside the workspace. Any version constraint you define will be ignored in this case. We recommend that you use ``"*"`` for source dependencies to make it clear that the version is inherited from the workspace.
```

Please check whether I've introduced any errors or fantasies as I've updated the text to make it less academic 🙈

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+When ``codeql pack install`` is executed from the query pack directory, an appropriate version of ``codeql/cpp-all`` will be downloaded to the local package cache. Also, a ``codeql-pack.lock.yml`` file will be created that contains the resolved version of ``codeql/cpp-all``. The lock file won't contain an entry for ``my-company/my-library`` since it is resolved from source. The ``codeql-pack.lock.yml`` file will look something like this:
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:76]:
```suggestion
When you execute ``codeql pack install`` from the query pack directory, an appropriate version of ``codeql/cpp-all`` is downloaded to the local package cache. Also, a ``codeql-pack.lock.yml`` file is created that contains the resolved version of ``codeql/cpp-all``. The lock file won't contain an entry for ``my-company/my-library`` since it is resolved from source dependencies. The ``codeql-pack.lock.yml`` file will look something like this:
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+.. code-block:: yaml
+
+   dependencies:
+     codeql/cpp-all:
+       version: 0.2.2
+
+When ``codeql pack publish`` is executed from the query pack directory, the ``codeql/cpp-all`` dependency from the package cache and the ``my-company/my-library`` from the workspace will be bundled with ``my-company/my-queries`` and published to the GitHub container registry.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:84]:
```suggestion
When you execute``codeql pack publish`` from the query pack directory, the ``codeql/cpp-all`` dependency from the package cache and the ``my-company/my-library`` from the workspace are bundled with ``my-company/my-queries`` and published to the GitHub container registry.
```

File: docs/codeql/codeql-cli/codeql-cli-reference.rst [hunk 15]
```diff
@@ -14,10 +14,12 @@ Learn more about the files you can use when running CodeQL processes and the res
...
    extractor-options
+   about-codeql-workspaces
 
 - :doc:`About CodeQL packs <about-codeql-packs>`: CodeQL packs are created with the CodeQL CLI and are used to create, depend on, publish, and run CodeQL queries, libraries, and query suites.
 - :doc:`Query reference files <query-reference-files>`: A query reference file is text file that defines the location of one query to test.
 - :doc:`SARIF output <sarif-output>`: CodeQL supports SARIF as an output format for sharing static analysis results.
 - :doc:`Exit codes <exit-codes>`: The CodeQL CLI reports the status of each command it runs as an exit code.
   This exit code provides information for subsequent commands or for other tools that rely on the CodeQL CLI.
 - :doc:`Extractor options <extractor-options>`: You can customize the behavior of extractors by setting options through the CodeQL CLI.
+- :doc:`About CodeQL workspaces <about-codeql-workspaces>`: CodeQL workspaces are used to group multiple CodeQL packs together.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/codeql-cli-reference.rst:25]: I'd be tempted to put the two references here to the new article immediately after the article about CodeQL packs, since they appear to be somewhat related.

**Review by felicitymay — COMMENTED** [review]
@aeisenberg - thanks for writing this new article about CodeQL workspaces - it was great learning about this 💖 

I've made a lot of small suggestions to update the language to make it more active and less academic in style. It's entirely possible that my suggestions would introduce errors or inaccuracies, so please do check carefully and revise or reject any that don't improve the clarity of the content.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+In most cases, the CodeQL workspace and all CodeQL packs contained in it should be stored in the same git repository so the development environment is more easily sharable.
+
+The ``codeql-workspae.yml`` file
+--------------------------------
+
+A CodeQL workspace is defined by a ``codeql-workspace.yml`` yaml file. This file contains a ``provide`` block, and optionally an ``ignore`` block. The ``provide`` block contains a list of glob patterns that define the CodeQL packs that are available in the workspace. The ``ignore`` block contains a list of glob patterns that define CodeQL packs that are not available in the workspace. Each entry in the ``provide`` or ``ignore`` section must map to a path to a ``qlpack.yml`` file. All glob patterns are relative to the directory containing the workspace file. See `@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>`__ for a list of patterns accepted in this file.
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst]: Looks like a duplicate comment. I will ignore.

[the remark this answers is not in the document]
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst]: Same.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+For example, the following ``codeql-workspace.yml`` file defines a workspace that contains all CodeQL packs recursively found in the ``codeql-packs`` directory, except for the packs in the ``experimental`` directory:
+
+.. code-block:: yaml
+
+    provide:
+      - "*/codeql-packs/**/qlpack.yml"
+    ignore:
+      - "*/codeql-packs/**/experimental/**/qlpack.yml"
+
+To verify that you have the correct ``codeql-workspace.yml`` file, run ``codeql pack ls`` command in the same directory as your workspace. The result of the command is a list of all CodeQL packs in the workspace.
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:33]: This is to check that your workspace file correctly references the packs that you are expecting. I will clarify.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst (hunk 11 above)
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:49]: Thanks for the suggestion. It's not quite right. And I can understand where your confusion is coming from.  Version constraints are defined in the qlpack.yml. 

Source dependencies are generally useful in two situations:

- The main use case is when you are doing local development and want to override some dependencies as you work out some changes across multiple packs during development.
- You want to override a dependency in your cache for some reason (maybe there's a bug in the published version, or you have made a custom tweak to one of the published packs).

I can clarify this a little better and incorporate parts of your suggestion.

Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:49]: Hmmm...thinking about this some more, it's even more complicated. You _can_ specify `--additional-packs` when you run `pack publish`. This means that you can publish packs using source resolution to override the default resolution. I thought we had disabled that, but we didn't. I can't see a situation where this would be a good thing to do (since you will be publishing your pack with contents that don't match the qlpack.yml).

I'll make sure to mention that doing so is not recommended.

Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:49]: To be clear, I was distinguishing source dependencies resolved from the same workspace (which 100% _should_ be in the published pack) with source dependencies resolved from `--additional-packs`, which probably shouldn't.

The reason is that with dependencies in the workspace, we expect that all packs are tightly connected and should change in lock-step with each other. Also, they should all be available in the same git repo, so it is easy to edit all of them at once.

Source dependencies from `--additional-packs` can come from anywhere and it may be hard to track down exactly which versions are being used.

However...it's also true that all published query packs contain all of their transitive dependencies, so even if you override some of the expected dependencies with source dependencies, you can inspect them in the published pack. 

Apologies for going into the weeds here. I'm trying to figure out what is important to mention and what is superfluous for the docs.

Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst]: @felicitymay I'm not sure if this paragraph answers your questions. 

https://github.com/github/codeql/pull/10182/commits/94a6506cbb6899c709a47edca0a2a5109d20dba4#diff-8c14f9abe9524e8201b042c7288e073c2db2fb3d7adb88a5bf79427879a018e9R49-R54

Please let me know if this is clear enough to you.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+.. include:: ../reusables/beta-note-package-management.rst
+
+CodeQL workspaces are used to group multiple CodeQL packs together. A typical use case for a CodeQL workspace is to develop a set of CodeQL library and query packs that are mutually dependent. For more information on CodeQL packs, see ":doc:`About CodeQL packs <about-codeql-packs>`."
+
+The main benefit of a CodeQL workspace is that it makes it easier for you to develop and maintain multiple CodeQL packs. When you use a CodeQL workspace, all the CodeQL packs in the workspace are available as *source dependencies* for each other when you run a CodeQL command that resolves queries. This makes it easier to develop, maintain, and publish multiple, related CodeQL packs.
+
+In most cases, you should store the CodeQL workspace and the CodeQL packs contained in it in one git repository. This makes it easier to share your CodeQL development environment.
+
+The ``codeql-workspace.yml`` file
+--------------------------------
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:15]: It looks as if the underline here is too short, causing the test failure that's blocking the Sphinx docs generation. Will attempt to fix so that I can review the output.

```suggestion
The ``codeql-workspace.yml`` file
----------------------------------
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+.. pull-quote::
+
+  Note
+
+  Source dependencies are CodeQL packs that are resolved from the local file system. These dependencies can be in the same CodeQL workspace, or specified as a path option using the ``--additional-packs`` argument. When you compile and run queries locally, source dependencies override any dependencies found in the local package cache as well as version constraints defined in the ``qlpack.yml``.
+
+  This is particularly useful in the following situations:
+
+  - One of the dependencies of the query pack you are running is not yet published. Resolving from source is the only way to refernce that pack.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:53]: Small typo:

```suggestion
  - One of the dependencies of the query pack you are running is not yet published. Resolving from source is the only way to reference that pack.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+.. code-block:: yaml
+
+    name: my-company/my-queries
+    version: 1.0.0
+    dependencies:
+      my-company/my-library: "*"
+      codeql/cpp-all: ~0.2.0
+
+Notice that the ``dependencies`` block for the CodeQL query pack,``my-company/my-queries``, specifies  ``"*"`` as the version of the library pack. Since the library pack is already defined as a source dependency in ``codeql-workspace.yml``, the library pack's content is always resolved from inside the workspace. Any version constraint you define will be ignored in this case. We recommend that you use ``"*"`` for source dependencies to make it clear that the version is inherited from the workspace.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:84]: Formatting issue in the generated content:

image

Likely fix.

```suggestion
Notice that the ``dependencies`` block for the CodeQL query pack, ``my-company/my-queries``, specifies  ``"*"`` as the version of the library pack. Since the library pack is already defined as a source dependency in ``codeql-workspace.yml``, the library pack's content is always resolved from inside the workspace. Any version constraint you define will be ignored in this case. We recommend that you use ``"*"`` for source dependencies to make it clear that the version is inherited from the workspace.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst (hunk 11 above)
Comment (felicitymay) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst]: Thanks. That looks clearer to me.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+
+The ``codeql-workspace.yml`` file
+----------------------------------
+
+A CodeQL workspace is defined by a ``codeql-workspace.yml`` yaml file. This file contains a ``provide`` block, and optionally an ``ignore`` block.
+
+* The ``provide`` block contains a list of glob patterns that define the CodeQL packs that are available in the workspace.
+* The ``ignore`` block contains a list of glob patterns that define CodeQL packs that are not available in the workspace.
+
+Each entry in the ``provide`` or ``ignore`` section must map to the location of a ``qlpack.yml`` file. All glob patterns are defined relative to the directory that contains the workspace file. For a list of patterns accepted in this file, see" `@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>`__ .
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:22]: Typo:

```suggestion
Each entry in the ``provide`` or ``ignore`` section must map to the location of a ``qlpack.yml`` file. All glob patterns are defined relative to the directory that contains the workspace file. For a list of patterns accepted in this file, see "`@actions/glob <https://github.com/actions/toolkit/tree/main/packages/glob#patterns>`__ ."
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+For example, the following ``codeql-workspace.yml`` file defines a workspace that contains all the CodeQL packs recursively found in the ``codeql-packs`` directory, except for the packs in the ``experimental`` directory:
+
+.. code-block:: yaml
+
+    provide:
+      - "*/codeql-packs/**/qlpack.yml"
+    ignore:
+      - "*/codeql-packs/**/experimental/**/qlpack.yml"
+
+To verify that your ``codeql-workspace.yml`` file includes the CodeQL packs that you expect, run ``codeql pack ls`` command in the same directory as your workspace. The result of the command is a list of all CodeQL packs in the workspace.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:33]:
```suggestion
To verify that your ``codeql-workspace.yml`` file includes the CodeQL packs that you expect, run the ``codeql pack ls`` command in the same directory as your workspace. The result of the command is a list of all CodeQL packs in the workspace.
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+
+CodeQL workspaces and query resolution
+--------------------------------------
+
+All CodeQL packs in a workspace are available as source dependencies for each other when you run any CodeQL command that resolves queries or packs. For example, when you run ``codeql pack install`` in a pack directory in a workspace, any dependency that can be found in the workspace will be instead of downloading that dependency to the package cache and adding it to the ``codeql-pack.lock.yml`` file. For more information, see `:ref:Adding and Installing Dependencies <adding-and-installing-dependencies>`__.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:39]: Missing word and formatting issue:

```suggestion
All CodeQL packs in a workspace are available as source dependencies for each other when you run any CodeQL command that resolves queries or packs. For example, when you run ``codeql pack install`` in a pack directory in a workspace, any dependency that can be found in the workspace will be used instead of downloading that dependency to the package cache and adding it to the ``codeql-pack.lock.yml`` file. For more information, see ":ref:`Adding and Installing Dependencies <adding-and-installing-dependencies>`."
```

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+
+When you execute ``codeql pack install`` from the query pack directory, an appropriate version of ``codeql/cpp-all`` is downloaded to the local package cache. Also, a ``codeql-pack.lock.yml`` file is created that contains the resolved version of ``codeql/cpp-all``. The lock file won't contain an entry for ``my-company/my-library`` since it is resolved from source dependencies. The ``codeql-pack.lock.yml`` file will look something like this:
+
+.. code-block:: yaml
+
+   dependencies:
+     codeql/cpp-all:
+       version: 0.2.2
+
+When you execute``codeql pack publish`` from the query pack directory, the ``codeql/cpp-all`` dependency from the package cache and the ``my-company/my-library`` from the workspace are bundled with ``my-company/my-queries`` and published to the GitHub container registry.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:94]: Space issue:

```suggestion
When you execute ``codeql pack publish`` from the query pack directory, the ``codeql/cpp-all`` dependency from the package cache and the ``my-company/my-library`` from the workspace are bundled with ``my-company/my-queries`` and published to the GitHub container registry.
```

**Review by felicitymay — COMMENTED** [review]
I found a few typos and formatting glitches. I'm going to merge those suggestions, so that I can regenerate the content.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+.. pull-quote::
+
+  Note
+
+  Source dependencies are CodeQL packs that are resolved from the local file system. These dependencies can be in the same CodeQL workspace, or specified as a path option using the ``--additional-packs`` argument. When you compile and run queries locally, source dependencies override any dependencies found in the local package cache as well as version constraints defined in the ``qlpack.yml``.
+
+  This is particularly useful in the following situations:
+
+  - One of the dependencies of the query pack you are running is not yet published. Resolving from source is the only way to reference that pack.
+  - You are making changes to multiple packs at the same time and want to test them together. Resolving from source ensures that you are using the version of the pack with your changes in it.
```
Comment (felicitymay) [inline docs/codeql/codeql-cli/about-codeql-workspaces.rst:54]: I should have spotted this in my original review. It seems as if this should be a section with its own title somewhere in this article. On line 10 we introduce the term in italics, but we don't currently have anyway to link to more information about source dependencies.

I'd recommend moving this information into a new section called "Source dependencies" or "Source dependencies in a workspace". Exactly where this section should go, and what other changes we should make to the rest of the content isn't entirely clear to me. It seems as if this concept is fundamental to the purpose of a workspace, so could arguably go towards the top of the article.

**Review by felicitymay — COMMENTED** [review]
Thanks for responding to my review comments. I've fixed the typos/formatting glitches introduced by the updates 🤦🏻‍♀️ 

I have one further comment about the way information is organized, but once that's resolved think that this is probably ready to publish.

File: docs/codeql/codeql-cli/about-codeql-workspaces.rst (hunk 25 above)
Comment (aeisenberg) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst:54]: Thank you. That is a good suggestion. I had a stab at this. I decided to put the source dependencies immediately after the section describing the `codeql-workspace.yml` file itself. I think this works ok.

Comment (felicitymay) [reply docs/codeql/codeql-cli/about-codeql-workspaces.rst]: 👍🏻 Thanks 🙂

**Review by felicitymay — APPROVED** [review]
